src/os/bluestore/HybridAllocator_impl.h

```diff
@@ -0,0 +1,186 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include "HybridAllocator.h"
+
+#define dout_context (T::get_context())
+#define dout_subsys ceph_subsys_bluestore
+#undef  dout_prefix
+#define dout_prefix *_dout << (std::string(this->get_type()) + "::").c_str()
+
+template <typename T>
+int64_t HybridAllocatorBase<T>::allocate(
+  uint64_t want,
+  uint64_t unit,
+  uint64_t max_alloc_size,
+  int64_t  hint,
+  PExtentVector* extents)
+{
+  dout(10) << __func__ << std::hex
+    << " 0x" << want
+    << "/" << unit
+    << "," << max_alloc_size
+    << "," << hint
+    << std::dec << dendl;
+  ceph_assert(std::has_single_bit(unit));
+  ceph_assert(want % unit == 0);
+
+  if (max_alloc_size == 0) {
+    max_alloc_size = want;
+  }
+  if (constexpr auto cap = std::numeric_limits<uint32_t>::max();
+    max_alloc_size >= cap) {
+    max_alloc_size = p2align(uint64_t(cap), (uint64_t)T::get_block_size());
+  }
+
+  std::lock_guard l(T::get_lock());
+
+  // try bitmap first to avoid unneeded contiguous extents split if
+  // desired amount is less than shortes range in AVL
+  bool primary_first = !(bmap_alloc &&
+                         bmap_alloc->get_free() &&
+                         want < T::_lowest_size_available());
+
+  int64_t res = __allocate_or_rollback(primary_first,
+    want, unit, max_alloc_size, hint, extents);
+  ceph_assert(res >= 0); // all the errors to be handled inside
+                         // __allocate_or_rollback()
+  if ((uint64_t)res < want) {
+    auto orig_size = extents->size();
+    int64_t res2 = 0;
+    // try alternate allocator
+    if (!primary_first) {
+      res2 = T::_allocate(want - res, unit, max_alloc_size, hint, extents);
+    } else if (bmap_alloc) {
+      res2 =
+        bmap_alloc->allocate(want - res, unit, max_alloc_size, hint, extents);
+    }
+    if (res2 >= 0) {
+      res += res2;
+    } else {
+      // allocator shouldn't return new extents on error
+      ceph_assert(orig_size == extents->size());
+    }
+  }
+  return res ? res : -ENOSPC;
+}
+
+template <typename T>
+void HybridAllocatorBase<T>::dump()
+{
+  std::lock_guard l(T::get_lock());
+  T::_dump();
+  if (bmap_alloc) {
+    bmap_alloc->dump();
+  }
+  dout(0) << __func__
+    << " avl_free: " << T::_get_free()
+    << " bmap_free: " << (bmap_alloc ? bmap_alloc->get_free() : 0)
+    << dendl;
+}
+
+template <typename T>
+void HybridAllocatorBase<T>::init_rm_free(uint64_t offset, uint64_t length)
+{
```

Comment:
In non-corrupted hybrid allocator state there is always a gap between chunk free in primary (avl/btree/btree2) and chunk free seconary (bitmap).
```
                  1                 2                 3                 4                 5
0        8       f0        8       f0        8       f0        8       f0        8       f0
rrrrrrrrr bbbbb bb bb b     b b rrrrrrrrrrrrrrrrrr       rrrrrrrrrrrrrrrrrrrr b rrrrrrrrrr
```
r - ranged, b - bitmap

It is not possible to have: `rrrrrrbbbbbb` because bitmap part will be moved to range,
and they would form continuous `rrrrrrrrrrrr`.


So, there are only 2 kinds of valid init_rm_free(off,len) ranges:
a) [off-len) is completely inside some range `rrrrrrrrr`
b) bitmap has all set in range [off-len) `bbbbbbb`

I claim that logic of _try_remove_from_tree can be much simpler.

Reply:
Likely you're right. But this is borrowed from the original implementation so I'd prefer to leave it as-is for now.